### PR TITLE
JSON IDL

### PR DESCRIPTION
**Pull request by Future-Outlier**
## Tracking issue

https://github.com/flyteorg/flyte/issues/5318

## Why are the changes needed?
1. solve more than 10+ issues due to the protobuf struct doesn't support int.
2. using msgpack bytes, which are smaller than JSON bytes, we can reduce storage size and improve I/O performance when interacting with database.

### flyteidl
- [x] add JSON IDL type

### flytepropeller
- [x] attribute access for remote execution
- [x] backward compatibility (compiler)
    upstreamType (Struct) -> downstreamType (Json)

### flytectl
- [x] flytectl create execution with inputs
- [x] flytectl get execution

## How was this patch tested?
unit tests and remote execution.
Example are from [here](https://github.com/flyteorg/flytekit/pull/2600)

####  [flytepropeller] attribute access
- dataclass example
image

#### [flytepropeller] backward compatible
- reference task 
upstream (struct) -> downstream (json)
image

- cache
upstream (struct) -> downstream (json)
image

#### [flytepropeller] others
dict example
image

#### [flytectl] create execution with inputs
```bash
(dev) future@outlier ~ % flytectl create execution --execFile build/PR/JSON/demo/flytectl/create_json_dataclass.yaml -p flytesnacks -d development

execution identifier project:"flytesnacks" domain:"development" name:"f6dee626da95445a8894"
```
image

```yaml
iamRoleARN: ""
inputs:
    input:
        a: 1
        b: 3.14
        c: example_string
        d:
            "1": 100
            "2": 200
        e:
            a: 1
            b: 3.14
envs: {}
kubeServiceAcct: ""
targetDomain: ""
targetProject: ""
task: dataclass_example.dataclass_task
version: x1QUSQo64Vd9FuIo7xBkvg
```

#### [flytectl] get task
```
(dev) future@outlier ~ % flytectl get task -d development -p flytesnacks dataclass_example.dataclass_task --execFile build/PR/JSON/demo/flytectl/get_json_dataclass.yaml --version x1QUSQo64Vd9FuIo7xBkvg
 ------------------------ ---------------------------------- ------------- -------- --------- -------------- ------------------- ----------------------------- 
| VERSION                | NAME                             | TYPE        | INPUTS | OUTPUTS | DISCOVERABLE | DISCOVERY VERSION | CREATED AT                  |
 ------------------------ ---------------------------------- ------------- -------- --------- -------------- ------------------- ----------------------------- 
| x1QUSQo64Vd9FuIo7xBkvg | dataclass_example.dataclass_task | python-task | input  | o0      |              |                   | 2024-08-26T12:27:05.188306Z |
 ------------------------ ---------------------------------- ------------- -------- --------- -------------- ------------------- ----------------------------- 
1 rows
```
```yaml
iamRoleARN: ""
inputs:
    input: []
envs: {}
kubeServiceAcct: ""
targetDomain: ""
targetProject: ""
task: dataclass_example.dataclass_task
version: x1QUSQo64Vd9FuIo7xBkvg
```
Note: For struct and JSON, the input will be an empty {} or []. This is an existing bug, and I am not planning to fix it in this PR. I genuinely believe this is a relatively low-priority issue."

## Check all the applicable boxes 

- [ ] I updated the documentation accordingly.
- [x] All new and existing tests passed.
- [x] All commits are signed-off.

### CODE REVIEW

File: flyteidl/clients/go/coreutils/extract_literal.go [hunk 1]
```diff
@@ -62,6 +62,8 @@ func ExtractFromLiteral(literal *core.Literal) (interface{}, error) {
 			return scalarValue.Generic, nil
 		case *core.Scalar_StructuredDataset:
 			return scalarValue.StructuredDataset.Uri, nil
+		case *core.Scalar_Json:
+			return scalarValue.Json.GetValue(), nil
```
Comment (Future-Outlier) [inline flyteidl/clients/go/coreutils/extract_literal.go:66]: This is covered by `flyteidl/clients/go/coreutils/literals.go`

File: flytepropeller/pkg/controller/nodes/attr_path_resolver_test.go [hunk 2]
```diff
@@ -328,3 +377,404 @@ func TestResolveAttrPathIn(t *testing.T) {
 		}
 	}
 }
+
+func TestResolveAttrPathInJson(t *testing.T) {
+	// Helper function to convert a map to JSON and then to msgpack
```
Comment (Future-Outlier) [inline flytepropeller/pkg/controller/nodes/attr_path_resolver_test.go:382]: 1. Test the map type first, then test the list type.
2. Within each category, order the tests by data type: integers (int), floats (float), strings (string), and nested maps or lists.
3. Finally, arrange the exception case tests.

File: flytepropeller/pkg/controller/nodes/attr_path_resolver.go [hunk 3]
```diff
@@ -79,34 +88,131 @@ func resolveAttrPathInPbStruct(nodeID string, st *structpb.Struct, bindAttrPath
...
+		if floatVal, err := vv.Float64(); err == nil {
+			return floatVal
+		}
+	}
+	return v
+}
+
 // convertInterfaceToLiteral converts the protobuf struct (e.g. dataclass) to literal
-func convertInterfaceToLiteral(nodeID string, obj interface{}) (*core.Literal, error) {
+func convertInterfaceToLiteral(nodeID string, obj interface{}, isJSON bool) (*core.Literal, error) {
```
Comment (pingsutw) [inline flytepropeller/pkg/controller/nodes/attr_path_resolver.go:174]: Could we add another function instead? `convertJsonToLiteral`
we can rename it to `convertStructToLiteral`

File: flyteidl/clients/go/coreutils/literals_test.go [hunk 4]
```diff
@@ -444,6 +455,47 @@ func TestMakeLiteralForType(t *testing.T) {
...
+			"b": 3.14,
+			"c": "example_string",
+			"d": map[string]interface{}{
+				"1": 100,
+				"2": 200,
+			},
+			"e": map[string]interface{}{
+				"a": 1,
+				"b": 3.14,
+			},
```
[the remark this answers is not in the document]
Comment (Future-Outlier) [reply flyteidl/clients/go/coreutils/literals_test.go:471]: yes, no problem!

File: flyteidl/protos/flyteidl/core/literals.proto [hunk 5]
```diff
@@ -46,6 +46,10 @@ message Binary {
     string tag = 2;
 }
 
+message Json {
```
Comment (pingsutw) [inline flyteidl/protos/flyteidl/core/literals.proto:49]: could we add some comments here

File: flytepropeller/pkg/controller/nodes/attr_path_resolver.go [hunk 6]
```diff
@@ -79,13 +88,93 @@ func resolveAttrPathInPbStruct(nodeID string, st *structpb.Struct, bindAttrPath
...
+	decoder := json.NewDecoder(strings.NewReader(jsonStr))
+	decoder.UseNumber()
+
+	err = decoder.Decode(&tmpVal)
+	if err != nil {
+		return nil, err
+	}
+	currVal = convertNumbers(tmpVal)
+
+	// Turn the current value to a map so it can be resolved more easily
```
Comment (pingsutw) [inline flytepropeller/pkg/controller/nodes/attr_path_resolver.go:122]: hmm, where did you turn the current value to a map?

File: flytepropeller/pkg/controller/nodes/attr_path_resolver.go [hunk 7]
```diff
@@ -106,7 +195,57 @@ func convertInterfaceToLiteral(nodeID string, obj interface{}) (*core.Literal, e
...
+		if err != nil {
+			return nil, err
+		}
+		literal.Value = scalar
+	}
+
+	return literal, nil
+}
+
+// convertJSONToLiteral converts the protobuf struct (e.g. dataclass) to literal
```
Comment (pingsutw) [inline flytepropeller/pkg/controller/nodes/attr_path_resolver.go:220]:
```suggestion
// convertJSONToLiteral converts the JSON (e.g. dataclass) to literal
```

**Comment by Future-Outlier** [comment]
@eapolinario  Can you review?
Thank you

File: flytepropeller/pkg/controller/nodes/attr_path_resolver.go (hunk 6 above)
Comment (Future-Outlier) [reply flytepropeller/pkg/controller/nodes/attr_path_resolver.go:122]: This is my fault, nice catch!!

File: flyteidl/protos/flyteidl/core/literals.proto (hunk 5 above)
Comment (Future-Outlier) [reply flyteidl/protos/flyteidl/core/literals.proto:49]: Yes of course!

**Review by eapolinario — COMMENTED** [review]
Can you comment on the use of github.com/vmihailenco/msgpack to handle msgpack-encoded values? I see that this is not listed in the official msgpack [website](https://msgpack.org/index.html).

File: flyteidl/protos/flyteidl/core/literals.proto [hunk 8]
```diff
@@ -46,6 +46,29 @@ message Binary {
...
+// deserialized into its corresponding structure.
+// This design ensures that the data is stored in a format that can be
+// fully reconstructed without loss of information.
+message Json {
+    // The JSON object serialized as a byte array.
+    bytes value = 1;
+
+    // The format used to serialize the byte array.
+    // This field identifies the specific format of the serialized JSON data,
+    // allowing future flexibility in supporting different JSON variants.
```
Comment (wild-endeavor) [inline flyteidl/protos/flyteidl/core/literals.proto:62]: add a comment here saying that serialization formats need to be supported in python, golang, rust, and javascript, in order for the full flyte experience to work.  (just throwing rust in there for now, we don't have anything yet).

Comment (Future-Outlier) [reply flyteidl/protos/flyteidl/core/literals.proto:62]: Yes, you're right, this is a pretty good and important advice, thank you.

Comment (wild-endeavor) [inline flyteidl/protos/flyteidl/core/literals.proto:62]: remove the json part of this comment, it's no longer true.

Comment (Future-Outlier) [reply flyteidl/protos/flyteidl/core/literals.proto:62]: yes, thank you.

**Comment by Future-Outlier** [comment]
Move to https://github.com/flyteorg/flyte/pull/5742

**Pull request closed without merging**